### PR TITLE
Add PUT Repository High Level REST API

### PR DESCRIPTION
**Pull request by hub-cap**
This commit adds PUT Repository, the associated docs and tests for the
high level REST API client. A few small changes to the PutRepository
Request and Response went into the commit as well.

Relates #27205

### CODE REVIEW

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java [hunk 1]
```diff
@@ -656,6 +657,18 @@ static Request getRepositories(GetRepositoriesRequest getRepositoriesRequest) {
...
 
+    static Request putRepository(PutRepositoryRequest putRepositoryRequest) throws IOException {
+        String endpoint = new EndpointBuilder().addPathPart("_snapshot").addPathPart(putRepositoryRequest.name()).build();
+        Request request = new Request(HttpPut.METHOD_NAME, endpoint);
+
+        Params parameters = new Params(request);
+        parameters.withMasterTimeout(putRepositoryRequest.masterNodeTimeout());
+        parameters.withTimeout(putRepositoryRequest.timeout());
+
+        request.setEntity(createEntity(putRepositoryRequest, REQUEST_BODY_CONTENT_TYPE));
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: I see in the corresponding REST action that we also support the `verify` parameter. I guess we should support it here too?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java [hunk 2]
```diff
@@ -67,4 +67,28 @@ public void getRepositoriesAsync(GetRepositoriesRequest getRepositoriesRequest,
...
             GetRepositoriesResponse::fromXContent, listener, emptySet(), headers);
     }
+
+    /**
+     * Creates a snapshot repository.
+     * <p>
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html"> Snapshot and Restore
+     * API on elastic.co</a>
+     */
+
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java]: remove this empty line?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 3]
```diff
@@ -1448,6 +1451,18 @@ public void testGetRepositories() {
         assertThat(expectedParams, equalTo(request.getParameters()));
     }
 
+    public void testPutRepository() throws IOException {
+        String repository = "repo";
+        PutRepositoryRequest putRepositoryRequest = new PutRepositoryRequest(repository);
+        putRepositoryRequest.type(FsRepository.TYPE);
+        putRepositoryRequest.settings("{\"location\": \".\"}", XContentType.JSON);
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: would it be possible to randomize the values that the request holds like in other tests?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryResponse.java [hunk 4]
```diff
@@ -49,4 +58,8 @@ public void writeTo(StreamOutput out) throws IOException {
         writeAcknowledged(out);
     }
 
+    public static PutRepositoryResponse fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryResponse.java]: can you add a test for this class although very simple?

File: server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java [hunk 5]
```diff
@@ -232,4 +233,17 @@ public void writeTo(StreamOutput out) throws IOException {
...
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field("name", name);
+        builder.field("type", type);
+        builder.startObject("settings");
+        settings.toXContent(builder, params);
+        builder.endObject();
+
+        builder.endObject();
+        return builder;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java]: can you add a unit test for this method?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java [hunk 6]
```diff
@@ -67,4 +67,28 @@ public void getRepositoriesAsync(GetRepositoriesRequest getRepositoriesRequest,
...
+    }
+
+    /**
+     * Asynchronously creates a snapshot repository.
+     * <p>
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html"> Snapshot and Restore
+     * API on elastic.co</a>
+     */
+    public void putRepositoryAsync(PutRepositoryRequest putRepositoryRequest,
+                                   ActionListener<PutRepositoryResponse> listener, Header... headers){
```
Comment (nik9000) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java]: Add a space after `)` and before `{` please!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java [hunk 7]
```diff
@@ -67,4 +67,28 @@ public void getRepositoriesAsync(GetRepositoriesRequest getRepositoriesRequest,
...
+            PutRepositoryResponse::fromXContent, emptySet(), headers);
+    }
+
+    /**
+     * Asynchronously creates a snapshot repository.
+     * <p>
+     * See <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-snapshots.html"> Snapshot and Restore
+     * API on elastic.co</a>
+     */
+    public void putRepositoryAsync(PutRepositoryRequest putRepositoryRequest,
```
Comment (javanna) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotClient.java]: in the spec we call this "snapshot.create_repository" . Could you rename the methods to createRepository please?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java [hunk 8]
```diff
@@ -1448,6 +1451,18 @@ public void testGetRepositories() {
         assertThat(expectedParams, equalTo(request.getParameters()));
     }
 
+    public void testPutRepository() throws IOException {
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: testCreateRepository?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SnapshotClientDocumentationIT.java [hunk 9]
```diff
@@ -58,7 +62,114 @@
  */
 public class SnapshotClientDocumentationIT extends ESRestHighLevelClientTestCase {
 
-    private static final String testRepository = "test_repository";
+    private static final String repositoryName = "test_repository";
+
+    public void testSnapshotPutRepository() throws IOException {
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SnapshotClientDocumentationIT.java]: testSnapshotCreateRepository?

File: client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SnapshotClientDocumentationIT.java [hunk 10]
```diff
@@ -58,7 +62,114 @@
...
+        PutRepositoryResponse response = client.snapshot().putRepository(request);
+        // end::put-repository-execute
+
+        // tag::put-repository-response
+        boolean acknowledged = response.isAcknowledged(); // <1>
+        // end::put-repository-response
+        assertTrue(acknowledged);
+    }
+
+    public void testSnapshotPutRepositoryAsync() throws InterruptedException {
```
Comment (javanna) [inline client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SnapshotClientDocumentationIT.java]: testSnapshotCreateRepositoryAsync

File: docs/java-rest/high-level/snapshot/put_repository.asciidoc [hunk 11]
```diff
@@ -0,0 +1,139 @@
+[[java-rest-high-snapshot-put-repository]]
+=== Snapshot Put Repository API
```
Comment (javanna) [inline docs/java-rest/high-level/snapshot/put_repository.asciidoc]: in the docs as well, can you rename put repository to create repository. We leave requests and responses as they are as we reuse them from the transport client, but for the rest we should follow the spec naming like other language clients already do.

**Review by nik9000 — APPROVED** [review]
I like it!

**Review by javanna — CHANGES_REQUESTED** [review]
thanks @hub-cap I left some comments!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java (hunk 1 above)
Comment (hub-cap) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/RequestConverters.java]: yea definitely, ty for pointing it out. Cant believe i missed it!

File: client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java (hunk 3 above)
Comment (javanna) [reply client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java]: I assumed that there is no problem setting values and checking that the output of the conversion from high-level request to low-level request is the expected one. We don't validate etc. I would do only what is straight-forward.

File: docs/java-rest/high-level/snapshot/put_repository.asciidoc (hunk 11 above)
Comment (hub-cap) [reply docs/java-rest/high-level/snapshot/put_repository.asciidoc]: I went ahead and changed it in all the headings and links (eg, `[[java-rest-high-snapshot-create-repository]]`)

**Comment by hub-cap** [comment]
@javanna ive addressed comments, pls review. I have also added repository request/response tests so i can cover most of the new things im adding to these request/responses in this and subsequent PRs.

File: server/src/test/java/org/elasticsearch/action/admin/cluster/repositories/RepositoryResponseTests.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import org.elasticsearch.action.admin.cluster.repositories.put.PutRepositoryResponse;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.json.JsonXContent;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class RepositoryResponseTests extends ESTestCase {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/repositories/RepositoryResponseTests.java]: could this extend AbstractXContentTestCase or AbstractStreamableXContentTestCase ?

**Review by javanna — APPROVED** [review]
thanks @hub-cap left one comment on testing, LGTM otherwise. feel free to push once addressed

**Comment by lcawl** [comment]
I think the docs/java-rest/high-level/snapshot/create_repository.asciidoc is missing from the 6.x branch. Is that intentional?  It's causing some documentation build issues.

**Comment by hub-cap** [comment]
@lcawl I had not yet pushed it. I was in backport hell for a while today. I pushed it just now.

**Pull request merged**